[TFL] Convert Size to Prod(Shape)

WindQAQ opened the pull request:
This PR converts `TF_SizeOp` to `TF_ProdOp(TF_ShapeOp)`. It makes TFL users can run models with `tf.size` without `SELECT_TF_OPS`. Fixes #46285.

nouiz committed:
Do not creates MaybeOwningDeviceMemory and ExecutionInput object in TF/XLA. This lower XLA launch overhead.

rkuester committed:
micro: copy operator FILL kernel from lite

This is a copy without modification of the kernel and test for
operator FILL from tensorflow/lite/kernels at 67993f46cb9.
Adaptations to micro and addition to the micro build to follow.

benbarsdell committed:
Move/enable helper functions in training_ops_gpu

- Moves the ROCm helper functions to the top of the file and enables
  them for use with CUDA as well.
- No functional change.

benbarsdell committed:
Add GPU kernels for SparseApply[Proximal]Adagrad

- Also applies to Resource and V2 versions of the ops.

rkuester committed:
micro: fix formatting in inactive copy of op FILL

fill_test.cc as coped from lite does not pass a formatting check
during CI. Even though this file is not a part of the build, it
must still pass a formatting check to be merged. In this case,
use clang-format to fix the order of the inclusions.

rkuester committed:
micro: replace ReportError() with TF_LITE_KERNEL_LOG macro

Replace naked calls to ReportError() with the TF_LITE_KERNEL_LOG
macro to satisfy an internal CI check, even though this file is
not presently included in the build.

ddavis-2015 committed:
Extract a function for parsing operator ADD_N

Extract the parsing out of a switch statement case to create a
standalone function which can be called by the micro op resolver.

PR step 1 for issue #46162

mansnils committed:
TFLu: Update downloaded GCC version to 10

* Add new GCC download script.
* Enable GCC 10 for cortex_m_generic and stm32f4 target.
* Update GCC flags for cortex_m_generic target.

rkuester committed:
micro: delete gmock/gtest includes in copy of FILL

Delete the gmock and gtest includes in the currently inactive
copy of op FILL. There is an annoying difference between internal
and external versions of the code that causes clang-formatting
errors in CI checks (see issue #46297). When the code is imported
internally, the gmock and gtest includes become:

    #include "/internal/path/to/gmock/gmock.h"

and then clang-format wants them to be in the lower include block
(i.e., next to "flatbuffers/flatbuffers.h").

The solution for now is to delete gmock and gtest includes when
making these operator copies.

advaitjain committed:
Have artifcat directory be different for different build_types.

Fixes #46261

mgalgs committed:
TFLite: iOS: Ensure we use the iOS mktemp

It's fairly common for iOS developers to have mktemp from coreutils
installed and sitting in front of the iOS mktemp on their PATH. Since the
coreutils version has different flags, trying to build tflite from source
results in the following error:

    ERROR: /Users/mgalgs/development/tensorflow/tensorflow/lite/ios/BUILD:49:28: Executing genrule //tensorflow/lite/ios:TensorFlowLiteC_framework failed (Exit 1): bash failed: error executing command /bin/bash -c ... (remaining 1 argument(s) skipped)
    mktemp: too few X's in template 'framework'
    Target //tensorflow/lite/ios:TensorFlowLiteC_framework failed to build

Fix this by using the mktemp from /usr/bin/ explicitly.

benbarsdell committed:
Set use_gpu=True in _testTypesForSparseAdagrad

advaitjain committed:
Allow more of the internal checks to have open-source counterparts.

 * gtest includes will be flagged as errors
 * use of the error reporter without the wrapper macros will be an error.
 * assert can not be used (static_assert is ok).

Manually tested by adding the disallowed strings to the code and
confirmed that an error is raised.

Fixes #46297
Fixes http://b/175657165

advaitjain committed:
Fix the checks to determine pass/fail for test_code_style.

WindQAQ committed:
Convert Size to Prod(Shape)

Add a test for unranked tensor

Add trailing endline

Convert Size to Prod(Shape)

WindQAQ committed:
Convert Size to Prod(Shape)

WindQAQ committed:
Convert Size to Prod(Shape)

Move the pattern to lower_tf

Use scalar for reduction_indices

Remove unranked size test

Constrain the lowering pattern to ranked tensor

majnemer committed:
Return an error instead of crashing

The function returns a Status so we can just propagate the stream error.

PiperOrigin-RevId: 351055448
Change-Id: Id7234429e063b2c6a590ac4ec36acd07a6992cea

abattery reviewed (commented):
Could you move this pattern into lower_tf.cc under tensorflow/compiler/mlir/tensorflow/transforms?

tensorflower-gardener committed:
Merge pull request #46271 from mgalgs:tflite-ios-build-mktemp

PiperOrigin-RevId: 351096399
Change-Id: Ifbed4b5242984b780cea73e811734c43c33ee7dd

mansnils committed:
TFLM: Set new GCC version to default

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2021-01-11

PiperOrigin-RevId: 351100322
Change-Id: I12add1b32814be44e930081ca3b19e152bc129f6

tensorflower-gardener committed:
Update GraphDef version to 643.

PiperOrigin-RevId: 351100327
Change-Id: Id62f84ea5c733599d7f125a8dbad0597bd73491d

tensorflower-gardener committed:
Merge pull request #45022 from nouiz:upstream_maybeowning

PiperOrigin-RevId: 351103206
Change-Id: Ifa9821e3dc1d4fd15c440da19a7e03d26daaa89e

Commit:
Fix iOS static framework link error by using custom object file extractor

With the bazel version update from 3.1.0 to 3.7.2, the object file hash values
are no longer added when their names are unique within an objc_library. See:
* https://github.com/bazelbuild/bazel/issues/11846
* https://github.com/bazelbuild/bazel/pull/11958

However, object name collision can still happen when there is a source code with
the same basename in a transitive dependency tree. Normally there is no problem
with this, but this had an unfortunate interaction with the symbol hiding script
we use for building the iOS static framework. That is, when an archive file (.a)
contains more than one object file with the same name, the 'ar -x' command
executed as part of the symbol hiding script would overwrite the conflicting
object file, causing the overwritten object file to be not included in the final
static framework.

This was causing a link error at CocoaPods lint step, as it had some missing
function definitions. This is now fixed by using a custom extraction script
written in Python, which gracefully handles the object file name collision.

This also fixes a previously existed race condition when the symbol hiding
script is run in parallel for multiple static framework targets, by using
separate temporary directories while extracting the object files.

Verified that this fix works with CocoaPods lint.

PiperOrigin-RevId: 351118550
Change-Id: Iec26e4720c21f271822785032d5fb6f4717eebca

tensorflower-gardener committed:
[MLIR][KernelGen] Add atan2 kernels

PiperOrigin-RevId: 351120975
Change-Id: I07bc4518bdaf66fcc4065c42a4cb393e742fba1e

tensorflower-gardener committed:
Merge pull request #45457 from rkuester:feature-micro-add-op-fill-pr3

PiperOrigin-RevId: 351140406
Change-Id: I75439803ef603e1ba22e6982762cc9f8d0a2b8f9

pifon2a committed:
[KERNEL_GEN] Add a canonicalization pattern to drop a redundant dynamic reshape.

PiperOrigin-RevId: 351141868
Change-Id: I7ef94dcd4f6b0be9899db0fbaea579bab8554421

tensorflower-gardener committed:
Merge pull request #46202 from ddavis-2015:AddN-pr1

PiperOrigin-RevId: 351141252
Change-Id: Ia57d74aa2b89c33dbdca208309cc8c5d39f3f3b8

mansnils committed:
TFLM: Fix appollo3evb target build

tensorflower-gardener committed:
Merge pull request #45818 from benbarsdell:gpu-SparseApplyProximalAdagrad-gpu-proper

PiperOrigin-RevId: 351148337
Change-Id: Ifa95dfde1376137c8de1c17b9763fddaaff4de5d

mansnils committed:
TFLM: Fix appollo3evb target build again

miaout17 committed:
NFC: Replace "toco" with "converter" in loggings.

The testing is for both MLIR-based and TOCO based converters.
Calling these "toco" is confusing

PiperOrigin-RevId: 351158814
Change-Id: If659d97a2b38feee980d9533db83e73b6798e2d1

karimnosseir committed:
[lite] Add new methods for get input/output tensors using signature input/output
names.
getInputTensorFromSignature/getOutputTensorFromSignature

Example:
Tensor inputTensor = interpreter.getInputTensorFromSignature("input", "my_method");
Tensor outputTensor = interpreter.getOutputTensorFromSignature("output", "my_method");
# Also you can pass null as methodName if the model has one signature.
Tensor inputTensor = interpreter.getInputTensorFromSignature("input", null);
PiperOrigin-RevId: 351159797
Change-Id: I4ad223d21ccd8e0582c4e545ef35721570c276b6

smit-hinsu committed:
Legalize Zeta and Polygamma ops with the tf2xla fallback

Auto-generated op definitions for Zeta and Polygamma ops

PiperOrigin-RevId: 351168224
Change-Id: Ic05964ddf83f1604f36c28bdc20ac1bc80e04b91

hawkinsp committed:
Add workaround for pybind11 bug that means we were issue poor errors in the case of Python integer -> C integer conversion overflow.

PiperOrigin-RevId: 351169598
Change-Id: Id1aad681e5a1a33697b6ed745b126705efba1857

tensorflower-gardener committed:
Add -march=native to kokoro tests.

PiperOrigin-RevId: 351177587
Change-Id: I9b995db41ae35e9768001fcb03ce58360e586ad5

pifon2a committed:
[KERNEL_GEN] Add a pattern for hlo.dyn_broadcast->linalg to enable is_inf kernel.

PiperOrigin-RevId: 351179620
Change-Id: I65f80d525ff65e307385804913f692cea4c3eab5

doehyun committed:
Use kUnknownDimFromConst for output_tensors_as_shapes for Pack op.

Without this, Const -1 input to Pack may be used in different ops,
where SymblociShapeManager may annotate unknown shapes, which can cause
wrong graph optimization (especially, common subgraph ellimination,
combined with constant folding).

PiperOrigin-RevId: 351181242
Change-Id: I5baea7108a97345e9e6cf977ed5fb507013bd15f

miaout17 committed:
Bugfix: TFLite `If` with dynamically allocated tensors.

Fixes a bug when branch subgraphs have dynamically allocated
inputs/outputs.

The newly added tests are failing without the fix and passing
with the fix.

PiperOrigin-RevId: 351181346
Change-Id: I5bdec6179c03157df9fd012018983384a6a2cb81

lamberta committed:
Update description of Eigen license.
Remove outdated references to (L)GPL license.

Fixes: https://github.com/tensorflow/tensorflow/issues/41739
PiperOrigin-RevId: 351183834
Change-Id: Idc56db6f53d3d8981172d9ffd49d4c09d09e687a

roserg committed:
Updatable uniforms done as generic arguments.
Removed std::vector<UniformBufferDescriptor> uniform_buffers from task descriptor.

PiperOrigin-RevId: 351184139
Change-Id: I3452939191822ba2102b17ab721ee691738652bd

hanhanW committed:
Upstream mhlo.dot lowering to Linalg to MHLO repo.

We prototyped the lowering from mhlo.dot to linalg.matmul in IREE. Since Linalg
now supports matmul in tensors world, we can move the lowering logic to tensors
world, and upstream to legalize_to_linalg.cc. The patch lowers the mhlo.dot to
the linalg.matmul/matvec/dot in tensors world.

PiperOrigin-RevId: 351184911
Change-Id: Ib43f501ca7ece18e7021c995733bc6d4b30d094e

tensorflower-gardener committed:
Extending the default PyModule_Type as a FastModuleType and use it for TFModuleWrapper. This extended type has improved attribute lookup mechanism to shorten tf api lookup speed.

PiperOrigin-RevId: 351187183
Change-Id: I02fe01dfff4f0669c7f5fbf0c1e7f3641a891e6d

fchollet committed:
Remove irrelevant warnings.
- tf.train is not longer part of the API.
- `Sequence`-based generator training is a LTS API and should not print warning when used the way it's intended to be used.

PiperOrigin-RevId: 351188267
Change-Id: Ibf8e47f4cc475351df732f91a8cb2105fc42ea13

fchollet committed:
Add smart_resize option to image_dataset_from_directory.

PiperOrigin-RevId: 351188455
Change-Id: Ifb24fabc529c266f5c99d9486f191b40ebfd76cf

tensorflower-gardener committed:
Replace the unused READY_WITH_RETENTION with the new VERY_LOW_POWER state.

PiperOrigin-RevId: 351189568
Change-Id: Icfa5b7ed3cd78fbbbb88552b4cab1efcbc525da2

dubey committed:
Disable flaky fault_tolerance_test on mac.

PiperOrigin-RevId: 351192115
Change-Id: I10e531d34680f1b06c86afd248f6dedfbd7c2934

roserg committed:
Storing tensor buffer handles in generic arguments.

PiperOrigin-RevId: 351192260
Change-Id: I3ab758f441bc89d5f957c65223d9d67580137f04

Commit:
Integrate LLVM at llvm/llvm-project@89878e8c966a

Updates LLVM usage to match
[89878e8c966a](https://github.com/llvm/llvm-project/commit/89878e8c966a)

PiperOrigin-RevId: 351192927
Change-Id: I7f1193cee52aaf997bd498db066e53da342c555f

qlzh727 committed:
Remove the instance check of _UnreadVariable (private tf api) in Keras.

PiperOrigin-RevId: 351194027
Change-Id: I373b1fe4df9f68d0ceb019f22222f05496a57375

skye committed:
[libtpu] Plumb HloModule::debug_options_ through the C API.

PiperOrigin-RevId: 351196886
Change-Id: Idfa6fe5a01969efb888cc9a22718e72d881f1835

Commit:
[XLA:SPMD] Adding support for detecting potential parallel dimensions in gather.

Some gathers have indices driven by IOTAs that cause the access to be
partitionable in the dimension indexed by the dimension with of the IOTA.
An example is:

  %iota.1 = iota()
  %indices = concatenate(..., %iota.1, ...)
  ... = gather(..., %indices)

which is a pattern followed by tf.reverse_sequence for example.

This patch adds support to detecting such situation and allowing sharding
of the gather across devices.

PiperOrigin-RevId: 351205227
Change-Id: I79b7b6bf9392a81df0f5810c41d4b72c6f2678a6

roserg committed:
Fusion of nodes simplified.
Logic of tasks fusion moved to compute_task_descriptor.

PiperOrigin-RevId: 351205320
Change-Id: Iaa632b1b6bb7bfc302debc831ac43b7f9c6a9d0b

WindQAQ committed:
Move the pattern to lower_tf

WindQAQ commented:
Done. Moved the pattern to tf-lower-tf.

tayo committed:
Internal change to BUILD.

PiperOrigin-RevId: 351216273
Change-Id: Ia5006772e3cd3535b23afabb8f8b94e24672971d

roserg committed:
Merging of nodes done as in OpenCL inference context.

PiperOrigin-RevId: 351217823
Change-Id: I00e795dc5bd0f2382e521799501ed23e2530a89a

jayxiaojieshi committed:
[tf.data] Rolls out the optimization `map_parallelization` as experiment to 5% of Borg jobs.

PiperOrigin-RevId: 351218900
Change-Id: Iad9c50673c49d6cd136ad7c1826ffe59035c877e

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -252,6 +258,17 @@ def : Pat<(TF_RsqrtGradOp $lhs, $rhs),
+
+// Size(x) = Prod(Shape(x), axis=0, keepdims=false)
+def : Pat<(TF_SizeOp:$result $arg),
+          (TF_ProdOp (CreateTFShapeOp $result, $arg, (IsI32 $result)),
```
/*reduction_indices=*/(GetScalarOfType<0> $result)

As reduction_indices in prod to be of i32 if the elements are i32 and I think scalar reduction indices are allowed.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:
```diff
@@ -252,6 +258,17 @@ def : Pat<(TF_RsqrtGradOp $lhs, $rhs),
+def : Pat<(TF_SizeOp:$result $arg),
+          (TF_ProdOp (CreateTFShapeOp $result, $arg, (IsI32 $result)),
+                     (TF_ConstOp (GetI64ScalarElementsAttr<0>)),
+                     ConstBoolAttrFalse)>;
```
nit: /*keep_dims=*/ConstBoolAttrFalse for readability.

smit-hinsu reviewed (commented):
Thanks for the contributions!

zhangyujing committed:
Fix a flaky test

PiperOrigin-RevId: 351221534
Change-Id: I9ce64bb360f9c1ca8de25af3b3207c9a3d38d0ce

tayo committed:
Internal change to BUILD.

PiperOrigin-RevId: 351224395
Change-Id: I14be33622b6e6ef097221cfc977e61b187b1ff3f

WindQAQ commented:
Hi @smit-hinsu, TF->TF lowering pass is also being used in `tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc`, so this PR will break `size_unranked` in `tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir`. Should I remove it or correct it to the right lowering? Thank you!

https://github.com/tensorflow/tensorflow/blob/009675faa9422ef8e2c8a30216b31adad62d25b6/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir#L3882-L3887

richard-uhler committed:
Verify resource subtype is supplied for tf.VarHandleOp.

If no resource subtype is supplied, we can't derive the dtype attribute.
Verify that a resource subtype is supplied so we get a verification failure
instead of an assert failure if you try to access the dtype attribute for a
tf.VarHandleOp of type tensor<*x!tf.resource>.

PiperOrigin-RevId: 351226121
Change-Id: I6505724067f410f4b2618a516c42293d0494e003

rxsang committed:
Improve docstring for `data.experimental.bucket_by_sequence_length` API.

PiperOrigin-RevId: 351226890
Change-Id: I01583d21b434a876d3a4b7fe11f5a091393ed5fa

smit-hinsu reviewed (commented):
I will need to take a closer look but this will also change lowering for the static shaped input which I think should be fine.

Could you try removing the lowering pattern as well?

tensorflower-gardener committed:
Merge pull request #46298 from advaitjain:style-check-improvements

PiperOrigin-RevId: 351229496
Change-Id: I6fe634c597e68529fde4306ba38545f61f4b1f7c

allenlavoie committed:
Stop running variable restoration twice when loading SavedModels

Even worse, the second time split the restore into one op per variable, which is very slow for high-latency filesystems.

This was an unfortunate side-effect of a 1.x compatibility case. Adds a unit test asserting that only one restore op is run.

PiperOrigin-RevId: 351231481
Change-Id: I43cb7a140f10ad743a64bd7f6893abc77477a90f

WindQAQ commented:
Sure, I can constrain the pattern to ranked input only.

Commit:
Add TpuExecutable_FreeXlaShapeIndexArray and TpuExecutable_FreeMaybeOwningDeviceMemoryArray to free appropriate memory

PiperOrigin-RevId: 351234388
Change-Id: I9ac9c712ad3a2e715d1a849bce7fbc6b4c858c43

k-w-w committed:
Remove unneeded lines.

There's no need to keep a separate set to track the seen nodes (`visited_nodes` because `path` is already effectively filtering out the duplicate nodes)

PiperOrigin-RevId: 351235385
Change-Id: I5004076b6b3e32d70261b3af04a4872236034887

WindQAQ committed:
Merge

WindQAQ committed:
Add a test for unranked tensor

WindQAQ committed:
Add trailing endline

smit-hinsu reviewed (commented):
Why not simply remove that pattern? Your new pattern should have the same behavior in case of static shapes.

Commit:
First attempt at a Python fuzzer.

Using the following as a guide:
https://google.github.io/oss-fuzz/getting-started/new-project-guide/python-lang/

PiperOrigin-RevId: 351239322
Change-Id: I91df8559d00f00e35ffdaec0960abe1b0a4462b1

WindQAQ commented:
> Why not simply remove that pattern? Your new pattern should have the same behavior in case of static shapes.

Ohoh, I misunderstand it :-(. Will do. Thank you!

WindQAQ committed:
Remove tf.Size legalization

tensorflower-gardener committed:
Merge pull request #46287 from mansnils:GCC10

PiperOrigin-RevId: 351240405
Change-Id: I981b38cb044d7b1fb26970cec0a96bc727fb12ba

tensorflower-gardener committed:
Merge pull request #46265 from advaitjain:fix-46261

PiperOrigin-RevId: 351241936
Change-Id: I856f62833ffcbe2792e6f91b7848ca1c90e9e3e7

WindQAQ committed:
Merge master

WindQAQ commented:
Umm, sorry that I mess up the branch. Let me reopen a new one.
